standardized-metadata-nested.py

```diff
@@ -0,0 +1,259 @@
+
+
+def build_metadata_dict(
+    namespaced_metadatas: Sequence[NamespacedMetadata], metadata: Mapping[str, Any]
+) -> Mapping[str, Any]:
+    namespaced_metadata_dict = {}
+    for namespaced_metadata in namespaced_metadatas:
+        namespaced_metadata_dict[namespaced_metadata.namespace()] = {
+            "_version": namespaced_metadata.version()
+        }
+        for key, value in namespaced_metadata.dict().items():
+            if value is not None:
+                namespaced_metadata_dict[namespaced_metadata.namespace()][key] = value
+
+    return {**metadata, "__namespaced_metadata": namespaced_metadata_dict}
+
+
+@asset
+def asset1():
+    return MaterializeResult(
+        metadata=build_metadata_dict(
+            namespaced_metadatas=[
+                AssetMetadata(
+                    storage_kind="snowflake", storage_address_string="my_db.my_schema.asset1"
+                ),
+                TableMaterializationMetadata(
+                    num_rows_total=500,
+                    num_rows_inserted=5,
+                    columns=[TableColumn(name="user_id", data_type="str")],
+                ),
+                SnowflakeTableMetadata(
+                    snowflake_address=SnowflakeTableAddress(
+                        database="my_db", db_schema="my_schema", table_name="asset1"
+                    )
+                ),
+            ],
+            metadata={
+                "my_unschematized_piece_of_metadata": 5,
+                "staging_address": SnowflakeTableAddress(
+                    database="my_staging_db", db_schema="my_schema", table_name="asset1"
+                ).dict(),
+            },
+        )
+    )
+
+
+####################################################################################################
+# An asset with metadata constructed using plain JSON-objects. Has identical metadata to the asset
+# above.
+####################################################################################################
+
+
+@asset
+def asset2():
+    return MaterializeResult(
+        metadata={
+            "__namespaced_metadata": {
+                "dagster.table": {
+                    "_version": "1.0.0",
+                    "num_rows_inserted": 5,
+                    "columns": [{"name": "user_id", "data_type": "str"}],
+                    "num_rows_total": 500,
+                },
+                "dagster": {
+                    "_version": "1.0.0",
+                    "storage_kind": "snowflake",
+                    "storage_address_string": "my_db.my_schema.asset1",
+                },
+                "dagster_snowflake": {
+                    "_version": "1.0.0",
+                    "snowflake_address": {
+                        "database": "my_db",
+                        "db_schema": "my_schema",
+                        "table_name": "asset1",
+                    },
+                },
+            },
```

Comment:
This makes a ton more sense to me. I can understand how this system works just by looking at this serialization format. 👍🏻